src/software/ai/navigator/obstacle/obstacle_factory_test.cpp

```diff
@@ -0,0 +1,173 @@
+    }
+    catch (std::bad_cast)
+    {
+        ADD_FAILURE() << "PolygonObstacle was not created for a rectangle";
+    }
+}
+
+TEST_F(ObstacleFactoryTest, create_ball_obstacle)
+{
+    Point origin(2.5, 4);
+    Rectangle rectangle(Point(1, 3), Point(5, 8));
+    Circle expected(origin, 0.1985);
+    Obstacle obstacle = obstacle_factory.createBallObstacle(origin);
+
+    try
+    {
+        auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
+        EXPECT_EQ(expected, circle_obstacle.getCircle());
+    }
+    catch (std::bad_cast)
+    {
+        ADD_FAILURE() << "CircleObstacle was not created for a ball";
+    }
+}
+
+TEST_F(ObstacleFactoryTest, create_robot_obstacle)
+{
+    Point origin(2.5, 4);
+    Rectangle rectangle(Point(1, 3), Point(5, 8));
+    Circle expected(origin, 0.207);
+    Obstacle obstacle = obstacle_factory.createRobotObstacle(origin);
+
+    try
+    {
+        auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
+        EXPECT_EQ(expected, circle_obstacle.getCircle());
+    }
+    catch (std::bad_cast)
+    {
+        ADD_FAILURE() << "CircleObstacle was not created for a robot";
+    }
+}
+
+TEST_F(ObstacleFactoryTest, stationary_robot_obstacle)
+{
+    Point origin(2.3, 3);
+    Vector velocity(0.0, 0.0);
+    Angle orientation(Angle::fromRadians(2.2));
+    AngularVelocity angular_velocity(AngularVelocity::fromRadians(-0.6));
+    Circle expected(origin, 0.041569219381653054);
+    Robot robot = Robot(3, origin, velocity, orientation, angular_velocity, current_time);
+    Obstacle obstacle = obstacle_factory.createVelocityObstacleFromRobot(robot);
+
+    try
+    {
+        auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
+        EXPECT_EQ(expected, circle_obstacle.getCircle());
+    }
+    catch (std::bad_cast)
+    {
+        ADD_FAILURE() << "CircleObstacle was not created for a stationary robot";
+    }
+}
+
+TEST_F(ObstacleFactoryTest, slow_moving_robot_obstacle)
+{
+    Point origin(-2.1, 5);
+    Vector velocity(0.0007, 0.004);
+    Angle orientation(Angle::fromRadians(2.2));
+    AngularVelocity angular_velocity(AngularVelocity::fromRadians(-0.6));
+    Circle expected(origin, 0.041569219381653054);
+    Robot robot = Robot(3, origin, velocity, orientation, angular_velocity, current_time);
+    Obstacle obstacle = obstacle_factory.createVelocityObstacleFromRobot(robot);
+
+    try
+    {
+        auto circle_obstacle = dynamic_cast<CircleObstacle&>(*obstacle);
+        EXPECT_EQ(expected, circle_obstacle.getCircle());
+    }
+    catch (std::bad_cast)
+    {
+        ADD_FAILURE() << "CircleObstacle was not created for a slow moving robot";
+    }
+}
+
+TEST_F(ObstacleFactoryTest, fast_moving_robot_obstacle)
+{
+    Point origin(-2.1, 5);
+    Vector velocity(0.27, 0.34);
+    Angle orientation(Angle::fromRadians(2.2));
+    AngularVelocity angular_velocity(AngularVelocity::fromRadians(-0.6));
+    Polygon expected({Point(-2.1325533066198088, 5.0258511552569072),
```

Comment:
These are super specific numbers that make me think this test is brittle. I suppose in theory it's fine since we're providing the config (see comment above) so only "true" implementation changes would change the results.

We don't have any "stretch" variables or other "high-level" metrics to compare polygons, and comparing points with a tolerance would only be marginally better. @garethellis0 @akhilveeraghanta thoughts on this? I'm undecided on how "bad" I think this is, just don't like seeing so many very precise numbers in a test.

Reply:
Hmm, for comparing individual points I think some distance tolerance is fine. Extending that to polygons, we really just want to get all the points from the polygon and check that they "match" our list of points to within some per-point distance tolerance.

Reply:
fixed that with equalWithinTolerance